[WIP] Jump-to-definition

j-f1 opened the pull request:
Ref #102

- [x] Tests
- [x] Simplified ids for top-level symbols (drop the index if `node.parent` is `SourceFile`?)

ry commented:
@j-f1 what's the status of this one?

j-f1 commented:
I want to finish it, just haven’t had a chance yet.

j-f1 commented:
I’ve used this scheme for the IDs:

- `export-${name}` or `export-default` for exports
- `symbol-${name}` for global symbols
- `symbol-${name}-${node.pos}` for local symbols. I’m willing to try to improve this one, but if you feel it’s a low priority, I’ll leave it as-is.

ry commented:
@j-f1 Wow very nice

I'm seeing some errors on http://localhost:4000/x/abc/deps.ts

<img width="1138" alt="Screen Shot 2019-08-12 at 5 12 27 PM" src="https://user-images.githubusercontent.com/80/62898844-791e1e00-bd24-11e9-83b7-9a1a6c7c0a36.png">

ry commented on src/analyze_code.js:
```diff
@@ -0,0 +1,281 @@
+ * @param  {ts.ImportDeclaration | ts.ExportDeclaration} statement
+ */
+function getRemoteLink(node, statement) {
+  return `${statement.moduleSpecifier.text}#export-${
```
Is it really necessary to prefix with `export-` ? I think it would be very nice to have a URL like this: https://deno.land/std@v0.9.0/fs/path/mod.ts#win32
to refer to the win32 export.

j-f1 replied:
Do you think it’s likely that there’d be conflicts? i.e. an export of a variable named `L1` would break jumping to lines.

ry replied:
Let's worry about conflicts later : )

j-f1 replied:
My worry is that changing anything about the way that IDs are generated after this PR is merged+deployed could lead to existing permalinks being broken.

ry replied:
I don't consider anchor links to be stable. Generally I consider this work to be a prototype which we will eventually use as a basis for a deno implementation.

I think we should aim for the beautiful interface and deal with the exceptional cases later. For example maybe we need to change line number links to be escaped. EG `#^L1` instead of `#L1`... But we can deal with that later.

j-f1 commented:
> I'm seeing some errors on http://localhost:4000/x/abc/deps.ts

Fixed! I was running the function to make links in comments & etc. actual links after I applied the transform that adds links to variables, causing the second linkifier to run on the HTML code itself.

I’ve also updated namespace imports (`foo` in `import * as foo from 'bar'`) to not link to an imaginary default export of the imported module.

ry commented:
I've deployed it to deno.land - looks great.

I have a few concerns - https://deno.land/std@v0.12/testing/asserts.ts#symbol-actual
<img width="476" alt="Screen Shot 2019-08-14 at 9 07 54 PM" src="https://user-images.githubusercontent.com/80/63066345-a2cb7680-bed7-11e9-95b5-fc434f9fd6a6.png">

What happens if there is another "actual" symbol as an arg to anther function. It seems to me that private symbols can just be assigned some numerical value for the anchor.

j-f1 commented:
Ooh, that’s a mistake! It should work like https://deno.land/std@v0.12/testing/asserts.ts#symbol-actualString-2333 further down.

j-f1 commented:
One thing that worries me is that the rendered code page takes a noticeable amount of time to load even on a fast connection.

ry commented:
@j-f1 Yeah... on deno.land it's behind CloudFront - so after the first load it seems okay. I currently have it setup to purge the entire cache on each update. We could be smarter about that. We could also attempt to pre-build the site before deployment... Your change is a great improvement, let's incur some technical debt and deal with the slowness later. I will land shortly.
